Passwordless-GDM: enable authentication method selection and EIdP login

ikerexxe opened the pull request:
This PR enables the authentication method selection and the EIdP login from the GNOME login screen.

The design page is available at https://github.com/SSSD/sssd.io/pull/79

The COPR repo for testing is available at https://copr.fedorainfracloud.org/coprs/ipedrosa/passwordles-gdm. You should update sssd, gnome-desktop3, mutter, gdm and gnome-shell packages.

In order to test it's necessary to set
```
[pam]
pam_json_services = gdm-switchable-auth
...
```

Constraints: if a user has enabled both password and EIdP authentication, they will only be able to authenticate using EIdP. There's a known problem with password authentication.

ikerexxe committed:
util: implement pam_get_response_data()

This API gets the selected response type data from the response_data
linked list. Includes unit tests.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>
Signed-off-by: Ray Strode <halfline@redhat.com>

ikerexxe committed:
Responder: generate JSON message for GUI

Implement a set of functions to check the available authentication
mechanisms and their associated data, and generate a JSON message with
it. This JSON formatted message will be consumed by apps that provide
GUI login (i.e. GDM). Currently, the implementation only takes into
account password and OAUTH2 mechanisms.

Include unit tests to check the implemented functions.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: call JSON message generation

Call JSON message generation function and fill the data structure
containing the response_data linked list.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
SSS_CLIENT: forward available auth JSON message

Forward the available authentication mechanisms and their associated
data message to the GUI login using a PAM conversation. Then, obtain the
reply and forward it to the responder, so that it can parse it.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>
Signed-off-by: Ray Strode <halfline@redhat.com>

ikerexxe committed:
Responder: unpack JSON reply from GUI

Implement a set of functions to unpack the JSON reply from the GUI.
Include unit tests to check the implemented functions.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: parse GUI reply

Parse GUI reply and set the appropriate data in `sss_auth_token`
structure.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: new option `pam_json_services`

This new option is used to enable the JSON protocol in the PAM responder
based on the PAM service file in use.

:config: Add pam_json_services option to enable JSON protocol to
         communicate the available authentication mechanisms.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Test: adapt test_pam_srv to JSON message

Include JSON message where applies.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Responder: check PAM service file for JSON protocol

Implement a function to check whether the PAM service file in use is
enabled for the JSON procotol. This helps us filter which applications
are compatible with this protocol.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe commented:
Remember to set
```
[pam]
pam_json_services = gdm-unified-auth
...
```
in sssd.conf in order to test this properly.

ikerexxe commented:
@justin-stephenson and @sumit-bose this is the PR for passwordless-GDM integration

justin-stephenson commented on src/man/sssd.conf.5.xml:
```diff
@@ -2062,6 +2062,29 @@ pam_gssapi_indicators_map = sudo:pkinit, sudo-i:pkinit
+                        <para>
+                            Comma separated list of PAM services where the JSON
+                            protocol to communicate the available
+                            authentication mechanisms is enabled.
```
Wording is a bit off IMO, maybe something like `Comma separated list of PAM services used to enable JSON authentication mechanism communication.`

or

`Comma separated list of PAM services used to communicate JSON authentication mechanism support.`

ikerexxe replied:
Updated the text with the first option. Thanks!

justin-stephenson reviewed (approved):
Ack, thank you.

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -0,0 +1,324 @@
+    }
+
+    ret = confdb_get_string(cdb, tmp_ctx, CONFDB_PC_CONF_ENTRY, CONFDB_PC_PASSWORD_PROMPT,
+                            "", &password_prompt);
```
Hi,

in general you should check the service-specific prompting option as well, i.e. there might be `[prompting/password/gdm-switchable-auth]` in `sssd.conf` which would overwrite settings from `[prompting/password].

But I think for better scalability it would be better if refactor  `pam_eval_prompting_config()` a bit so that there is a function which returns the `struct prompt_config` list and either access it directly or add getter-functions to return the configured prompting for the individual methods. I think the latter would be best because it would allow to have all the details about handling the prompting configuration in  `pam_prompting_config.c`.

Additionally, it might be worth config option the the oath2 `init_prompt` and `link_prompt` as well.

bye,
Sumit

ikerexxe replied:
I refactored ` pam_eval_prompting_config()` to return the `prompt_config` structure and created a new function called `pam_get_prompting_option()` to get the prompting.

On top of that, I added the `init_prompt` and `link_prompt` options for EIdP method.

sumit-bose commented on src/responder/pam/pamsrv_json.c:
```diff
@@ -322,3 +322,85 @@ generate_json_auth_message(struct confdb_ctx *cdb, struct pam_data *_pd)
+    }
+
+    ret = json_unpack(jroot, "{s:{s:s,s:o}}", "auth-selection",
+                      "status", &status, "password", &json_data);
```
Hi,

maybe using `json_object_foreach()` would help to avoid lots of code duplication when more methods are added?

bye,
Sumit

ikerexxe replied:
I updated it. Right now the function it's a little bit longer, but it'll help me when I add smartcard and passkey methods.

sumit-bose commented on src/man/sssd.conf.5.xml:
```diff
@@ -2062,6 +2062,28 @@ pam_gssapi_indicators_map = sudo:pkinit, sudo-i:pkinit
+                    <listitem>
+                        <para>
+                            Comma separated list of PAM services used to enable
+                            JSON authentication mechanism communication.
```
Hi,

I would suggest to change this sentence to something like `Comma separated list of PAM services which can handle the JSON protocol for selecting authentication mechanisms, see https://sssd.io/..... for details`.

bye,
Sumit

ikerexxe replied:
I changed the wording for the message except for the URL, as I don't think we are adding them anywhere and they might change so it's a little bit risky.

sumit-bose replied:
Thanks

sumit-bose commented on src/responder/pam/pamsrv_cmd.c:
```diff
@@ -377,6 +377,11 @@ static int pam_parse_in_data_v2(struct pam_data *pd,
                                            body, blen, &c);
                     if (ret != EOK) return ret;
                     break;
+                case SSS_PAM_ITEM_JSON_AUTH_SELECTED:
```
Hi,

I think `SSS_PAM_ITEM_JSON_AUTH_SELECTED` and the current `SSS_PAM_ITEM_AUTHTOK` should be mutually exclusive and if both are present it should be a fatal error.

bye,
Sumit

ikerexxe replied:
Updated

sumit-bose commented:
Hi,

thanks for the patch, after realizing that Fedora 40 already has some newer package version than your test repo everything was working fine in my tests. I also rebased the patches on top of current master to see if the change in krb5_child about handling multiple authentication methods will cause issues, but currently they don't.

The comments I added can be handled when moving forward adding more methods.

bye,
Sumit

ikerexxe committed:
Responder: tune prompts in the GUI

Return `prompt_config` structure  in `pam_eval_prompting_config` to tune
the prompts from the SSSD config in the GUI.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
sss_client: add EIdP to prompt_config structure

Integration with GDM requests two prompts for EIdP so adding them to
prompt_config structure. In addition, implement all the functions needed
to manipulate the structure for these new prompts. Finally, add
unit-tests for the new functions.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe committed:
Add new options to tune EIdP prompting

These new options are needed by the GDM integration, but they can be
reused for CLI prompting.

:config: New options to tune EIdP prompting: 'init_prompt' and
         'link_prompt'.

Signed-off-by: Iker Pedrosa <ipedrosa@redhat.com>

ikerexxe commented:
Hi @sumit-bose ,

This is ready for review again. CI failures are unrelated and [COPR](https://copr.fedorainfracloud.org/coprs/ipedrosa/passwordles-gdm/builds/) repo is updated with latest code, including prompt tuning.

sumit-bose commented on src/responder/pam/pam_prompting_config.c:
```diff
@@ -346,10 +347,67 @@ errno_t pam_eval_prompting_config(struct pam_ctx *pctx, struct pam_data *pd)
 done:
     free(resp_data);
-    pc_list_free(pc_list);
+
```
Hi,

I think in case of an error `pc_list` is leaking. I would suggest to add something like
```
    if (ret != EOK) {
        pc_list_free(pc_list);
    }
````

bye,
Sumit

ikerexxe replied:
Yeah, you are right. This is fixed

sumit-bose replied:
Thanks.

sumit-bose commented on src/responder/pam/pamsrv.h:
```diff
@@ -172,7 +175,24 @@ errno_t filter_responses(struct pam_ctx *pctx,
+ *         error code otherwise.
+ */
+errno_t pam_get_prompting_option(struct prompt_config **pc_list, int prompt_type,
+                                 int index, const char **_prompt);
```
Hi,

there already are
```
const char *pc_get_password_prompt(struct prompt_config *pc);
const char *pc_get_2fa_1st_prompt(struct prompt_config *pc);
const char *pc_get_2fa_2nd_prompt(struct prompt_config *pc);
const char *pc_get_2fa_single_prompt(struct prompt_config *pc);
const char *pc_get_passkey_inter_prompt(struct prompt_config *pc);
const char *pc_get_passkey_touch_prompt(struct prompt_config *pc);
```
in the client code. Imo it would be better to add `pc_get_eidp_init_prompt()` and `pc_get_eidp_link_prompt()` and reuse those functions in the PAM responder?

bye,
Sumit

ikerexxe replied:
Makes sense, so I updated the implementation to take this into account.

sumit-bose replied:
Thanks.

sumit-bose commented:
Hi,

thanks for the update, code looks good. I will run some tests, does the build in your copr repo contain the latest changes or would it be better if I build SSSD from this PR?

bye,
Sumit

ikerexxe commented:
You can test it using the COPR repository, as it has been updated.

sumit-bose commented:
Hi,

thanks, test were working fine. I have no further comments.

bye,
Sumit

ikerexxe commented:
@sumit-bose  can you approve it?

@justin-stephenson I made some changes after Sumit's review. Do you want to take another look?

sumit-bose reviewed (approved):
Hi,

all is looking and working fine, ACK.

bye,
Sumit

justin-stephenson commented:
> 
> @justin-stephenson I made some changes after Sumit's review. Do you want to take another look?

It's okay for me, Ack.

ikerexxe commented:
Thanks for the reviews. I'm closing this PR and rebasing the development branch manually to add the `Reviewed-by` tags.
